app/controllers/api/messaging.php

```diff
@@ -0,0 +1,2138 @@
+
+        $queueForDeletes
+            ->setType(DELETE_TYPE_SUBSCRIBERS)
+            ->setDocument($topic);
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_NOCONTENT)
+            ->noContent();
+    });
+
+App::post('/v1/messaging/topics/:topicId/subscribers')
+    ->desc('Adds a Subscriber to a Topic.')
+    ->groups(['api', 'messaging'])
+    ->label('audits.event', 'subscribers.create')
+    ->label('audits.resource', 'subscribers/{response.$id}')
+    ->label('scope', 'subscribers.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_JWT, APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'createSubscriber')
+    ->label('sdk.description', '/docs/references/messaging/create-subscriber.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_CREATED)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_SUBSCRIBER)
+    ->param('subscriberId', '', new CustomId(), 'Subscriber ID. Choose a custom Topic ID or a new Topic ID.')
+    ->param('topicId', '', new UID(), 'Topic ID.')
+    ->param('targetId', '', new UID(), 'Target ID.')
+    ->inject('dbForProject')
+    ->inject('response')
+    ->action(function (string $subscriberId, string $topicId, string $targetId, Database $dbForProject, Response $response) {
+        $subscriberId = $subscriberId == 'unique()' ? ID::unique() : $subscriberId;
+
+        $topic = Authorization::skip(fn () => $dbForProject->getDocument('topics', $topicId));
+
+        if ($topic->isEmpty()) {
+            throw new Exception(Exception::TOPIC_NOT_FOUND);
+        }
+
+        $target = Authorization::skip(fn () => $dbForProject->getDocument('targets', $targetId));
+
+        if ($target->isEmpty()) {
+            throw new Exception(Exception::USER_TARGET_NOT_FOUND);
+        }
+
+        $subscriber = new Document([
+            '$id' => $subscriberId,
+            '$permissions' => [
+                Permission::read(Role::user($target->getAttribute('userId'))),
+                Permission::delete(Role::user($target->getAttribute('userId'))),
+            ],
+            'topicId' => $topicId,
+            'topicInternalId' => $topic->getInternalId(),
+            'targetId' => $targetId,
+            'targetInternalId' => $target->getInternalId(),
+        ]);
+
+        try {
+            $subscriber = $dbForProject->createDocument('subscribers', $subscriber);
+            $dbForProject->deleteCachedDocument('topics', $topicId);
+        } catch (DuplicateException) {
+            throw new Exception(Exception::SUBSCRIBER_ALREADY_EXISTS);
+        }
+
+        $response
+            ->setStatusCode(Response::STATUS_CODE_CREATED)
+            ->dynamic($subscriber, Response::MODEL_SUBSCRIBER);
+    });
+
+App::get('/v1/messaging/topics/:topicId/subscribers')
+    ->desc('List topic\'s subscribers.')
```

Comment:
will a developer need to be able to search for subscribers across multiple topics?

Reply:
I can't think of a use case, can you?

Reply:
I can imagine a developer needing to determine what topics a user is part of.

In general, I'm against nesting resources because it makes it harder to query for data.

Reply:
Subscribers will have a userId, and the userName will be dynamically included, the same as we have for memberships of a team

Reply:
> Subscribers will have a userId, and the userName will be dynamically included, the same as we have for memberships of a team

This only helps the front end show the user name w/o an extra API call. Is there any other approach to finding which topics a user is part of without making an API call per topic?

Reply:
Hmm no, I see the difficulty, how do we handle this for teams? Is there a way to list teams a user is a member of?